### PR TITLE
Add evaluator toturial

### PR DESCRIPTION
**Pull request by thinksoso**
如何在Libai增加evaluator

### CODE REVIEW

File: docs/source/tutorials/Add Evaluator.md [hunk 1]
```diff
@@ -0,0 +1,67 @@
+# Add Evaluator
+## Evaluation
+Evaluation is a process that takes a number of inputs/outputs pairs and aggregate them. You can always use the model directly and just parse its inputs/outputs manually to perform evaluation. Alternatively, evaluation is implemented in LiBai using the `DatasetEvaluator` interface.
```
Comment (kaijieshi7) [inline docs/source/tutorials/Add Evaluator.md:3]: and aggregates

File: docs/source/index.rst [hunk 2]
```diff
@@ -15,6 +15,7 @@ Welcome to libai's documentation!
    tutorials/Getting Started.md
    tutorials/Train and Eval Command Line.md
    tutorials/Model Zoo.md
+   tutorials/Add Evaluator.md
```
Comment (rentainhe) [inline docs/source/index.rst:18]: 直接命名为Evaluation.md就行了

File: docs/source/tutorials/Add Evaluator.md [hunk 3]
```diff
@@ -0,0 +1,67 @@
+# Add Evaluator
```
Comment (rentainhe) [inline docs/source/tutorials/Add Evaluator.md:1]: 这里用`Evaluation`, 然后下面一个用`Evaluator`, 主要可以参考 https://detectron2.readthedocs.io/en/latest/tutorials/evaluation.html

File: docs/source/index.rst (hunk 2 above)
Comment (rentainhe) [reply docs/source/index.rst:18]: 奥这个文件名其实没关系，主要是`.md`里第一个标题要用`Evaluation`比较好

**Comment by rentainhe** [comment]
可以refer一下d2的文档或者是参考一些他里面的表达~

File: docs/source/tutorials/Evaluation.md [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+  def reset(self):
+    self.count = 0
+  def process(self, inputs, outputs):
+    for output in outputs:
+      self.count += len(output["instances"])
+  def evaluate(self):
+    # save self.count somewhere, or print it, or return it.
+    return {"count": self.count}
+```
+## Use evaluators
```
Comment (rentainhe) [inline docs/source/tutorials/Evaluation.md:16]: `Use evaluators` -> `Evaluator Usage`

File: docs/source/tutorials/Evaluation.md [hunk 5]
```diff
@@ -0,0 +1,87 @@
+# Evaluation
+Evaluation is a process that takes a number of inputs/outputs pairs and aggregates them. You can always use the model directly and just parse its inputs/outputs manually to perform evaluation. Alternatively, evaluation is implemented in LiBai using the `DatasetEvaluator` interface.
+
+LiBai includes a few `DatasetEvaluator` that computes metrics like Acc@N, PPL. You can also implement your own `DatasetEvaluator` that performs some other jobs using the inputs/outputs pairs. For example, to count how many instances are detected on the validation set:
```
[the remark this answers is not in the document]
Comment (rentainhe) [reply docs/source/tutorials/Evaluation.md:4]: For example, to count many instances are detected on the validation set，这句太明显了
不如直接refer to [d2 tutorials]()

File: docs/source/tutorials/Evaluation.md [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+class MyEvaluator(DatasetEvaluator):
+    def __init__(self):
+        self._predictions = []
+
+    def reset(self):
+        self._predictions = []
+
+    def process(self, inputs, outputs):
+        pred_logits = outputs["prediction_scores"]
+        labels = inputs["labels"]
```
Comment (CPFLAME) [inline docs/source/tutorials/Evaluation.md:54]: 我觉得这里要特别说明一下,  用户怎么得到 这个字典里面的`key`
outputs是个字典,  它来源于model.forward()返回的字典输出
inputs也是个字典, 它来源于trainer.get_data()里面对`dataset.__getitem__()`的输出进行转换后 得到的字典输出.

File: docs/source/tutorials/Evaluation.md [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+  def process(self, inputs, outputs):
+    for output in outputs:
+      self.count += len(output["instances"])
+  def evaluate(self):
+    # save self.count somewhere, or print it, or return it.
+    return {"count": self.count}
+```
+## Evaluator Usage
+To evaluate using the methods of evaluators manually:
+```
```
Comment (L1aoXingyu) [inline docs/source/tutorials/Evaluation.md:18]: ` ```python`

File: docs/source/tutorials/Evaluation.md [hunk 8]
```diff
@@ -0,0 +1,87 @@
...
+eval_results = inference_on_dataset(
+    model,
+    data_loader,
+    evaluator,
+    ...
+)
+```
+## Customize Evaluator using DatasetEvaluator
+`DatasetEvaluator` is the Base class for a dataset evaluator. This class will accumulate information of the inputs/outputs (by `process`), and produce evaluation results in the end (by `evaluate`).
```
Comment (CPFLAME) [inline docs/source/tutorials/Evaluation.md:39]:
```suggestion
`DatasetEvaluator` is the Base class for a dataset evaluator. This class will accumulate information of the inputs/outputs (by `process`) after every batch inference, and produce evaluation results in the end (by `evaluate`).
```

File: docs/source/tutorials/Evaluation.md [hunk 9]
```diff
@@ -0,0 +1,87 @@
...
+
+Firstly, create a new file in `libai/evaluation/`, and declare a new evaluator class that inherits the `DatasetEvaluator` and overwrites its `process` and `evaluation` functions to satisfy the needs.
```
Comment (L1aoXingyu) [inline docs/source/tutorials/Evaluation.md:41]: 这里不是一定要在 `libai/evaluation/` 里面去创建这个文件

File: docs/source/tutorials/Evaluation.md [hunk 10]
```diff
@@ -0,0 +1,87 @@
...
+            all_sample += len(preds)
+        self._results = OrderedDict()
+        self._results["acc"] = correct/all_sample
+        return copy.deepcopy(self._results)
+```
+
+Secondly, import the customized class in `libai/evaluation/__init__.py`
+``` Python
+from from .myevaluator import MyEvaluator
+```
```
Comment (L1aoXingyu) [inline docs/source/tutorials/Evaluation.md:78]: 这一步也不用，只需要最后在 config 里面 import 进来就可以了

File: docs/source/tutorials/Evaluation.md [hunk 11]
```diff
@@ -0,0 +1,87 @@
+# Evaluation
+Evaluation is a process that takes a number of inputs/outputs pairs and aggregates them. You can always use the model directly and just parse its inputs/outputs manually to perform evaluation. Alternatively, evaluation is implemented in LiBai using the `DatasetEvaluator` interface.
```
[the remark this answers is not in the document]
Comment (thinksoso) [reply docs/source/tutorials/Evaluation.md:2]: alternatively 参考的d2表达，意思应该是LiBai没有直接采用手动处理输入输出的方式，而是用了DatasetEvaluator接口

File: docs/source/tutorials/Evaluation.md [hunk 12]
```diff
@@ -0,0 +1,87 @@
...
+from from .myevaluator import MyEvaluator
+```
+Lastly, set the evaluation in config.
+``` Python
+from libai.evaluation import MyEvaluator
+evaluation=dict(
+    enabled=True,
+    evaluator=LazyCall(MyEvaluator)(),  # calculate top-k acc
+    eval_metric="acc",
+)
```
Comment (CPFLAME) [inline docs/source/tutorials/Evaluation.md:86]:
```suggestion
  evaluation=dict(
        enabled=True,
        # evaluator for calculating top-k acc
        evaluator=LazyCall(MyEvaluator)(),
        eval_period=5000,
        eval_iter=1e9,  # running steps for validation/test

        # Metrics to be used for best model checkpoint.
        eval_metric="acc", # your returned metric key in MyEvaluator.evaluate()
        eval_mode="max", # set `max` or `min` for saving best model according to your metric
    ),
)
```

File: docs/source/tutorials/basics/Evaluation.md [hunk 13]
```diff
@@ -0,0 +1,89 @@
...
+``` Python
+eval_results = inference_on_dataset(
+    model,
+    data_loader,
+    evaluator,
+    ...
+)
+```
+## Customize Evaluator using DatasetEvaluator
+`DatasetEvaluator` is the Base class for a dataset evaluator. This class will accumulate information of the inputs/outputs (by `process`) after every batch inference, and produce evaluation results in the end (by `evaluate`). The inputs is from the `trainer.get_data()`, which converts the outputs of `dataset.__getitem__()` to dict. The Outputs is from the dict return of `model.forward()`.
```
Comment (L1aoXingyu) [inline docs/source/tutorials/basics/Evaluation.md:39]: trainer.get_batch()

**Review by CPFLAME — APPROVED** [review]

**Review by rentainhe — APPROVED** [review]

File: docs/source/index.rst [hunk 14]
```diff
@@ -15,6 +15,7 @@ Welcome to libai's documentation!
    tutorials/Getting Started.md
    tutorials/Train and Eval Command Line.md
    tutorials/Model Zoo.md
+   tutorials/Evaluation.md
```
[the remark this answers is not in the document]
Comment (rentainhe) [reply docs/source/index.rst]: 额当我没说，这个是很早的review了，现在已经改完了

[the remark this answers is not in the document]
Comment (thinksoso) [reply docs/source/index.rst]: 嗯嗯，rst结构变了，这个文件就没了
